Unified GLES3 / GLES2 Batching

lawnjelly opened the pull request:
Very much a WIP, the new unified batching both provides acceleration for GLES3 and will be adding new features which will also improve GLES2 batching.

There will probably be issues with custom shaders to sort out, but it seems to run the platformer demo fine and a few test games and the editor.

Uses the curiously recurring template pattern to share code between the 2 backends.

Fixes #19943.
Fixes #37077.
Fixes #33225.
Fixes #42576.
Fixes #35772.
Fixes #36749.
Fixes #41324.
Mitigates #40905.

Benchmarking with the text benchmark I used for GLES2, I get 45x increase in fps, versus ~78x for GLES2. That is to be expected as the bottlenecks are slightly different in GLES3. This benchmark is specifically designed to look at drawcalls, real life increases are likely to be far less dramatic.

## Expectations
* As with GLES2 batching, any increases in performance will depend on how much your project is bottlenecked by drawcalls. Projects using tilemaps / lots of text will typically benefit the most. Custom shaders may or may not be able to be batched.
* Normal mapping angles are now hopefully fixed, although regressions are expected. I have merged PR #41323 into the batching.
* Projects with lights, you may get a slowdown due to fill rate. Be sure to try the `batching/light_scissoring` option.
* Batching settings are the same as for GLES2.
* Speed may not be optimal yet, at first I will be trying to fix any visual regressions.
* If you get decreased speed with batching, this is highly likely to be because batching defaults to `single_rect_fallback` to false. This is the old method of drawing quads which is about 2x as fast, but causes flicker on nvidia drivers. To directly compare the performance of both you would need to turn this back on.

## New features
This isn't just a straight port of the feature set of GLES2 batching, I am now adding new features, which will be available in GLES2 and GLES3.

* Light angles - for correct normal mapping angles and flips - DONE
* Large vertex format - for accelerating more custom shaders - DONE - needs testing - build 0.13
* More primitives - Ninepatch rects, lines, polys DONE
* Software skinning path for 2d polys - DONE
_(software skinning is switchable in projectsettings->rendering->quality->2d)_
* Switchable ninepatch margin scaling method - DONE (except GLES3 legacy)
_(switchable in projectsettings->rendering->quality->2d)_

## Please read for more info
https://docs.godotengine.org/en/3.2/tutorials/optimization/batching.html#doc-batching
https://docs.godotengine.org/en/3.2/tutorials/optimization/batching.html#frequently-asked-questions

## Builds
In progress builds for linux / windows. I'll try and update these as I go.
Please post any feedback regarding regressions, performance gains (or not) etc, with the build version you tried.

To find regressions, run while turning on `flash_batching` in project settings. The batching settings are exactly the same as GLES2. To test performance, turn off vsync, and choose to `debug/settings/stdout/print_fps` and run with batching switched off, then on.

* 0.16 Fixed software skinning, ninepatch options
* 0.15 Batching polys and software skinning path for polys
* 0.14 Batching ninepatch rects and lines (single pixel width only)
* 0.13 New vertex formats for accelerating custom shaders - highly likely to have bugs
* 0.12 Light angles for normal mapping
* 0.11 batching with items as well as commands - more likely to have regressions
* 0.10 First version - batching with commands only

https://github.com/lawnjelly/Misc/releases

lawnjelly commented:
@clayjohn I've changed this from draft to ready for review.

There's probably quite a few things I've missed as far as naming etc but I'm keen for us to get this in as soon as possible, even if more work is expected. This is partly so we can get testing asap (as Akien said he is building 3.2.4 betas soon) but also because this is such a major change that it involves substantial work to manually rebase every time any other PR is added to the rendering.

Some things of note:
* The common template to do the batching is quite large. For this reason and to prevent it becoming unwieldy I have split it into several `inc` files which are included from the main header. This is a trade off. It makes it super easy to navigate and find areas of the template concerned with different topics. On the other hand it breaks a lot of intellisense which is a pain.

I am open to moving them all into the same header, or a single hander and single inc, and separating the implementations and definitions if that is preferable. Or alternatively we could do such changes later if we thought it would work better in the long run.

* There's a really tricky bug in the joining I haven't figured out yet which is meaning the first in a series of items often isn't getting joined. This doesn't result in visual errors, but it would be nice to fix because it could potentially lead to better performance.
* I finally got around to solving the 'dummy' batches at the start of runs. This probably doesn't make that much difference but makes debugging and diagnostics easier.

clayjohn commented:
I've quickly reviewed the easy stuff, but Im gonna need a few more days to get through the difficult parts. So far it looks pretty good. Personally, I really like have things split across many files (even though some of those files are only 20 lines long). However, I know reduz really prefers long files that include all the needed functionality, so he will have to weigh in here. Either way it won't be too bad as most of those ``.inc`` files could be joined together without harming readability too much.

I'll let you know when I have a chance to do a more substantive review. The overall structure looks pretty similar to GLES2 batching, but I haven't really dug into how you are batching the other primitive types. So we'll see.

lawnjelly commented:
For batching the other types, the main conceptual change is that for the buffers, the start vertex and number of vertices (or commands) is stored in terms of vertex rather than quad (group of 4 verts) in the old version. This is because lines use 2 verts each, and polys can use a variable number of verts per poly.

What simplifies it considerably is that only similar primitives can be joined together at the moment. E.g. a group of rects are joined, and rendered as a group of rects. Poly group all polys etc.

This is something I'm meaning to improve later (probably after 3.2.4) so multiple types can be included in the same vertex buffer before a flush, it requires an extra layer of logic on top, but should result in a lot less VB uploads and probably better performance.

## Ninepatches
Ninepatches are very simple, in that they are just translated into 9 rects and drawn as rects. In the version you have it only supports the GLES2 method, stretch mode, but I've been starting to look into whether it is possible to support the tiling modes as in GLES3. This seems to be a whole can o' worms, not in terms of technical difficulty, but in terms of agreement about how ninepatches should work.

I've opened up a new proposal to discuss this, so this PR doesn't become too off topic with what is really a side issue:
https://github.com/godotengine/godot-proposals/issues/1618

lawnjelly commented:
As mentioned in the linked proposal, in unified batching path, ninepatch scaling mode is now selectable between a default old style (pixel perfect) and puthre's scaling version. These only support the stretch mode.

In GLES3 if the tile mode is switched on, or batching is off, it uses the shader path. To keep the shader simple, based on agreement it is reverted to the old style (it is kind of messy to support both in the shader). This does have the consequence that if you are using tiling in GLES3, only old style pixel perfect scaling is available. Based on feedback this seems the more popular version.

avencherus commented:
Ran some quick tests this morning comparing results using the benchmark project from:  https://github.com/godotengine/godot/issues/19943

[polygon2d_performance.zip](https://github.com/godotengine/godot/files/5362925/polygon2d_performance.zip)

Mostly gains for the Poly2D, but there was some regression in performance for Sprites in GLES3.

![image](https://user-images.githubusercontent.com/13004169/95707366-fcf1d400-0c61-11eb-93b6-9df941f162af.png)


**Source screenshots of the project tests:**

### Old Version

**GLES2 Sprite Tests**
![GLES2-s1-a](https://user-images.githubusercontent.com/13004169/95707458-44786000-0c62-11eb-801b-b00614d50048.png)
![GLES2-s2-a](https://user-images.githubusercontent.com/13004169/95707461-45a98d00-0c62-11eb-8160-029801469a17.png)
![GLES2-s3-a](https://user-images.githubusercontent.com/13004169/95707464-480be700-0c62-11eb-8e4f-cf94c248e239.png)

**GLES3 Sprite Tests**
![GLES3-s1-a](https://user-images.githubusercontent.com/13004169/95707471-4c380480-0c62-11eb-85af-0e2149cfaf95.png)
![GLES3-s2-a](https://user-images.githubusercontent.com/13004169/95707479-4fcb8b80-0c62-11eb-9192-b60de3ade23d.png)
![GLES3-s3-a](https://user-images.githubusercontent.com/13004169/95707483-50fcb880-0c62-11eb-9141-d1d40fa4d74c.png)

**GLES2 Poly Tests**
![GLES2-p1-a](https://user-images.githubusercontent.com/13004169/95707487-5528d600-0c62-11eb-93b4-911e23fcbb2b.png)
![GLES2-p2-a](https://user-images.githubusercontent.com/13004169/95707489-56f29980-0c62-11eb-9639-b51968922799.png)
![GLES2-p3-a](https://user-images.githubusercontent.com/13004169/95707491-5823c680-0c62-11eb-8458-1c16ac663a5c.png)

**GLES3 Poly Tests**
![GLES3-p1-a](https://user-images.githubusercontent.com/13004169/95707493-5b1eb700-0c62-11eb-8cff-cf02b4d4c47a.png)
![GLES3-p2-a](https://user-images.githubusercontent.com/13004169/95707495-5ce87a80-0c62-11eb-9d71-ed19774eb723.png)
![GLES3-p3-a](https://user-images.githubusercontent.com/13004169/95707500-607c0180-0c62-11eb-806e-57ba807f6ad3.png)


### New Version

**GLES2 Sprite Tests**
![GLES2-s1-b](https://user-images.githubusercontent.com/13004169/95707545-7f7a9380-0c62-11eb-9260-fee85c0df856.png)
![GLES2-s2-b](https://user-images.githubusercontent.com/13004169/95707546-80abc080-0c62-11eb-9c34-11d7c5322c20.png)
![GLES2-s3-b](https://user-images.githubusercontent.com/13004169/95707549-81dced80-0c62-11eb-9c70-5d6783759e87.png)

**GLES3 Sprite Tests**
![GLES3-s1-b](https://user-images.githubusercontent.com/13004169/95707574-8c978280-0c62-11eb-8cda-9fb10e0d92f2.png)
![GLES3-s2-b](https://user-images.githubusercontent.com/13004169/95707578-8e614600-0c62-11eb-965e-55741d997d27.png)
![GLES3-s3-b](https://user-images.githubusercontent.com/13004169/95707579-8f927300-0c62-11eb-8913-23d9ca4c1864.png)

**GLES2 Poly Tests**
![GLES2-p1-b](https://user-images.githubusercontent.com/13004169/95707534-7984b280-0c62-11eb-89c6-cbe09f65d3c5.png)
![GLES2-p2-b](https://user-images.githubusercontent.com/13004169/95707537-7ab5df80-0c62-11eb-93e8-4870248474c7.png)
![GLES2-p3-b](https://user-images.githubusercontent.com/13004169/95707540-7c7fa300-0c62-11eb-9ec7-794c01f6354c.png)

**GLES3 Poly Tests**
![GLES3-p1-b](https://user-images.githubusercontent.com/13004169/95707558-85707480-0c62-11eb-8b44-3f0b0c15c1cb.png)
![GLES3-p2-b](https://user-images.githubusercontent.com/13004169/95707562-873a3800-0c62-11eb-99ee-9bfa0259fad8.png)
![GLES3-p3-b](https://user-images.githubusercontent.com/13004169/95707565-8903fb80-0c62-11eb-86d0-eb4801761dae.png)

lawnjelly commented:
That all looks good. The drop in sprite performance depending on your test is expected .. as with the GLES2 batching the default single rect draw method is now equivalent to nvidia workaround, which is slower, but free of flicker artifacts.

If you want to directly compare sprite performance, you should set `project_settings->rendering->batching->options->single_rect_fallback` to true.

This has been a source of confusion in the past. I did actually change this default temporarily for some of the test builds, but when we integrate in 3.2 the default will be the same as in GLES2 (as it uses the same setting).

We really are trying to discourage people from using the old faster single rect method. The reason is, that while it may look great on your machine, a significant portion of your users may see flicker, and uninstall your game / give bad reviews.

lawnjelly committed:
Unified GLES2 / GLES3 Batching

Batching is mostly separated into a common template which can be used with multiple backends (GLES2 and GLES3 here). Only necessary specifics are in the backend files.

Batching is extended to cover more primitives.

lawnjelly commented:
Now rebased with the buffer orphaning.

I've also tidied up `canvas_batcher` a little bit, moved the implementations of most of the functions outside the template class so the declarations are easier to navigate. It is still a bit unwieldy to my eyes having so much in a monolithic class, but the templated nature makes it a little more difficult to split apart unless we use multiple templates, and the boiler plate could get excessive.

On boiler plate I've used macros to reduce this.

I've also changed a couple of the function names to slightly more sensible versions, let me know any more suggestions for renaming.

clayjohn commented:
Thanks for keeping it updated! I swear I will find time to review it soon. I've already read through all the code bits at a time, I just need to dedicate a chunk of time to do a more careful review of the substance.

clayjohn commented on drivers/gles2/rasterizer_canvas_gles2.cpp:
```diff
@@ -45,304 +45,160 @@ static const GLenum gl_primitive[] = {
+	// could these have ifs?
+	glDisableVertexAttribArray(VS::ARRAY_TEX_UV);
+	glDisableVertexAttribArray(VS::ARRAY_COLOR);
+	//	glDisableVertexAttribArray(VS::ARRAY_TANGENT);
```
cleanup

clayjohn commented on drivers/gles2/rasterizer_canvas_gles2.cpp:
```diff
@@ -45,304 +45,160 @@ static const GLenum gl_primitive[] = {
-				// unoptimized - could this be done once per batch / batch texture?
-				bool send_light_angles = rect->normal_map != RID();
+	// color
+	//	if (!colored_verts) {
```
cleanup

clayjohn commented on drivers/gles3/rasterizer_canvas_gles3.h:
```diff
@@ -28,136 +28,50 @@
 
-#ifndef RASTERIZERCANVASGLES3_H
-#define RASTERIZERCANVASGLES3_H
+#pragma once
```
@akien-mga are we starting to use #pragma once now?

lawnjelly replied:
Ah that's probably me from initial implementation, I'll remove the #pragma once. :+1:

clayjohn reviewed (changes requested):
I've gone through he whole thing, given the length of the PR, I can't say I spent a significant amount of time on each line of code. So I will mostly make general comments here:

1. I like the overall structure of the implementation. It is a nice logical step up from the previous batching implementation. 

2. The ``PREAMBLE`` macro was a great way to simplify the templates.

3. There is a lot of commented out code that has been left in throughout, I stopped commenting on them part way through. Before merging we will have to do a full read-through and make sure that we get read of the dead code.

Overall, great work! I am amazed at the size and complexity of this PR.

With just a little bit of code cleanup I think this will be ready for a testing build, we should do something similar to when we first added batching. @akien-mga can we get you to help put together another test batching build so we can get this into the hands of as many users as possible?

lawnjelly commented on drivers/gles2/rasterizer_canvas_gles2.cpp:
```diff
@@ -45,304 +45,160 @@ static const GLenum gl_primitive[] = {
 
-		RasterizerStorageGLES2::Texture *texture = storage->texture_owner.getornull(p_texture);
+	// orphan the old (for now)
+	//glBufferData(GL_ARRAY_BUFFER, 0, 0, GL_DYNAMIC_DRAW);
```
I actually changed this just yesterday, commenting out the glBufferCall. I didn't think it was necessary, and the test versions should reveal any problems (I think I had initially put it in the first batching for safety).

Every upload uses a glBufferData call rather than glSubBufferData, so will be requesting a new buffer each time. The buffer sizes are usually quite small, so at the moment this seems more appropriate than trying to maintain a large number of larger fixed sized buffers per frame. (This is same mechanism as GLES2 batching).

lawnjelly commented:
I've cleaned up a number of the comments... :+1: 

Some I did miss but part of the reason for leaving others in is that we are expecting to have to tweak things as a result of the betas, and the commented code may help for changing these areas (and some are placeholders). Bear in mind we can also do a separate cleanup PR after the betas.

lawnjelly committed:
Reduce glBufferSubData calls in legacy renderer

This is part of effort to make more efficient use of the API for devices with poor drivers. This eliminates multiple calls to glBufferSubData per update.

lawnjelly commented:
Added a second commit, doing some experiments to try and lower the amount of glBufferSubData calls in the legacy renderer in the hope of helping performance on devices with poor drivers.

akien-mga reviewed (approved):
Amazing work as usual!

Let's get this merged and tested in 3.2.4 beta 1 (hopefully sometime this week).

akien-mga commented:
Thanks!

@lawnjelly Do you want to prepare an issue with some testing instructions like we had for GLES2 batching, to collect user feedback?

lawnjelly commented:
Yup sure will do. :+1:

Valeryn4 commented:
The problem is still relevant.
The error pops up on some devices.

Version 3.3
![image](https://user-images.githubusercontent.com/12483173/133201566-f0fd1af8-821e-4330-b1e5-5c7a7686ef28.png)
![image](https://user-images.githubusercontent.com/12483173/133201512-d7ff138c-b4cf-4695-af8f-3595d356aa87.png)

Calinou commented:
@Valeryn4 Please continue the discussion about this in https://github.com/godotengine/godot/issues/35772.

oeleo1 commented:
@lawnjelly Not sure where to write this quick note, so here :-) The `RasterizerCanvasGLES2/3::try_join_item()` function seems to have a useless `reclip` var and a dead code test on it later on which is probably a leftover from previous versions of the code that you may want to cleanup.
